docs/hooks.rst

```diff
@@ -4,44 +4,177 @@
+
+The next code snippet demonstrates a basic usage of a `beforeSave` hook. This one
+will update field values just before record is saved::
 
     $m->addHook('beforeSave', function($m) {
-        $this['name'] = 'John';  // WRONG!
-        $m['surname'] = 'Smith'; // GOOD
+        $m['name'] = strtoupper($m['name']);
+        $m['surname'] = strtoupper($m['surname']);
     });
 
-    $m->insert([]);
-    // Will save into DB:  ['surname'=>'Smith'];
+    $m->insert(['name'=>'John', 'surname'=>'Smith']);
+
+    // Will save into DB:  ['name'=>'JOHN', 'surname'=>'SMITH'];
+
+Arguments
+---------
+
+When you define a callback, then you'll receive reference to model from all the hooks.
+It's important that you use this argument instead of $this to perform operation, otherwise
+you can run into problems with cloned models.
 
-    echo $m['name'];   // Will contain 'John', which it shouldn't
-                       // because insert() is not supposed to affect active record
+Callbacks does non expect anything to be returned, but you can modify fields of the
+model.
 
-afterLoad hook
---------------
-You can return false from afterLoad hook to prevent yielding of particular data rows.
+Interrupting
+------------
 
-Use it like this::
+You can also break all "before" hooks which will result in cancellation of the original action::
 
-$model->addHook('afterLoad', function ($m) {
-    if ($m['date'] < $m->date_from) {
-        $m->breakHook(false); // will not yield such data row
-    }
-    // otherwise yields data row
-});
+    $m->breakHook(false);
 
-Also this approach can be used to prevent data row to be loaded. If you return false
+If you break beforeSave, then the save operation will not take place, although model will assume
+the operation was successful.
+
+You can also berak beforeLoad hook which can be used to skip rows::
+
+    $model->addHook('afterLoad', function ($m) {
+        if ($m['date'] < $m->date_from) {
+            $m->breakHook(false); // will not yield such data row
+        }
+        // otherwise yields data row
+    });
+
+This will also prevent data from being loaded. If you return false
 from afterLoad hook, then record which we just loaded will be instantly unloaded.
-This can be helpful in some cases.
+This can be helpful in some cases, although you should still use :php:meth:`Model::addCondition`
+where possible as it is much more efficient.
+
+Insert/Update Hooks
+-------------------
+
+Insert/Update are triggered from inside save() method but are based on current state of
+:php:meth:`Model::loaded`:
+
+ - beforeInsert($m, &$data) (creating new records only)
+ - afterInsert($m, $id)
+ - beforeUpdate($m, &$data) (updating existing records only. Not executed if model is not dirty)
+ - afterUpdate($m)
+
+The $data argument will contain array of actual data (field=>value) to be saved, which you can
+use to withdraw certain fields from actually being saved into the database (by unsetting it's value).
+
```

Comment:
Tiny real-life example here would be good. I mean - when to use beforeSave and when to use beforeInsert/Update.

Reply:
added link to SoftDelete example here, which uses the technique.